### PR TITLE
Split RevaultTransaction's signature_hash() method

### PR DESCRIPTION
**Pull request by darosior**
I'm not that happy with the solution but it does really help as a caller.

### CODE REVIEW

File: src/transactions.rs [hunk 1]
```diff
@@ -109,92 +138,94 @@ pub trait RevaultTransaction: fmt::Debug + Clone + PartialEq {
...
-                    "No previous witness txo for psbtin: '{:?}'",
-                    psbtin
-                ))
-            })?;
-            if psbtin.non_witness_utxo.is_some() {
-                debug_assert!(
-                    false,
-                    "Cannot be reached. We never create transactions with non_witness_utxo."
-                );
+        let psbtin = match self.inner_tx_mut().inputs.get_mut(input_index) {
```
Comment (edouardparis) [inline src/transactions.rs:141]: use again ok_or 
```suggestion
        let psbtin = match self.inner_tx_mut().inputs.get_mut(input_index).ok_or(
           Error::InputStatisfaction(...
        )?;
```

File: src/transactions.rs [hunk 2]
```diff
@@ -109,92 +138,94 @@ pub trait RevaultTransaction: fmt::Debug + Clone + PartialEq {
...
-                        "Unknown expected sighash type for psbtin: '{:?}'",
-                        psbtin
-                    )));
-                }
-            };
-            if sighash_type != expected_sighash_type {
+        // -- If a sighash type is provided, the signer must check that the sighash is acceptable.
+        // If unacceptable, they must fail.
+        let (sig, sighash_type) = signature;
+        let expected_sighash_type = match psbtin.sighash_type {
```
Comment (edouardparis) [inline src/transactions.rs:205]: ok_or_else

File: src/transactions.rs [hunk 3]
```diff
@@ -71,30 +71,59 @@ pub trait RevaultTransaction: fmt::Debug + Clone + PartialEq {
...
+    fn signature_hash_internal_input(
+        &self,
+        input_index: usize,
+        sighash_type: SigHashType,
+    ) -> Option<SigHash> {
+        let psbt = self.inner_tx();
+        // TODO: maybe cache the cache at some point (for huge spend txs)
+        let mut cache = SigHashCache::new(&psbt.global.unsigned_tx);
+
+        if let Some(psbtin) = psbt.inputs.get(input_index) {
```
Comment (edouardparis) [inline src/transactions.rs:87]:
```suggestion
        psbt.inputs.get(input_index).map(|psbtin| {
           ...
        })
```

File: src/transactions.rs [hunk 4]
```diff
@@ -71,30 +71,59 @@ pub trait RevaultTransaction: fmt::Debug + Clone + PartialEq {
...
-            .and_then(|psbtin| psbtin.witness_utxo.as_ref())
-            .ok_or_else(|| {
-                Error::InputSatisfaction(format!(
-                    "Input index {} is out of bonds or psbt input has no witness utxo",
-                    input_index
-                ))
-            })?;
 
-        Ok(cache.signature_hash(input_index, &script_code, prev_txo.value, sighash_type))
+        if let Some(psbtin) = psbt.inputs.get(input_index) {
```
Comment (edouardparis) [inline src/transactions.rs:118]:
```suggestion
        psbt.inputs.get(input_index).map(|psbtin| {...})
```

File: src/transactions.rs (hunk 1 above)
Comment (darosior) [reply src/transactions.rs:141]: Just made it an assert!() right now as i'm going to tackle deserialization checks ASAP

Comment (darosior) [reply src/transactions.rs:141]: Just made it an assert!() right now as i'm going to tackle deserialization checks ASAP

File: src/transactions.rs (hunk 3 above)
Comment (darosior) [reply src/transactions.rs:87]: Applied

**Comment by darosior** [comment]
Thanks for the review. I've applied some of your suggestions and modified the first commit to outright panic as i expected to do Soon :tm:, but i'm now going to tackle https://github.com/re-vault/revault_tx/issues/40 ASAP so that our assumptions hold also for parsed PSBTs..

**Review by edouardparis — APPROVED** [review]
ACK 391afb351d9863c525fc556954577211b38ebced

**Pull request merged**